File: 21. Merge Two Sorted Lists Easy/solution.ts
class ListNode {
    val: number;
    next: ListNode | null;

    constructor(val: number, next: ListNode | null = null) {
        this.val = val;
        this.next = next;
    }
}

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const dummy: ListNode = new ListNode();
    let current: ListNode | null = dummy;

    while (list1 !== null && list2 !== null) {
        if (list1.val < list2.val) {
            current.next = list1;
            list1 = list1.next;
        } else {
            current.next = list2;
            list2 = list2.next;
        }
        current = current.next;
    }

    if (list1 !== null) {
        current.next = list1;
    }

    if (list2 !== null) {
        current.next = list2;
    }

    return dummy.next;
}

function printList(node: ListNode | null): void {
    while (node !== null) {
        console.log(node.val);
        node = node.next;
    }
}

const list1: ListNode = new ListNode(1, new ListNode(2, new ListNode(4)));
const list2: ListNode = new ListNode(1, new ListNode(3, new ListNode(4)));

const mergedList: ListNode | null = mergeTwoLists(list1, list2);
printList(mergedList);

File: 21. Merge Two Sorted Lists Easy/solution.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val=val
        self.next=next

def merge_two_lists(list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    dummy = ListNode()
    current = dummy
    
    while list1 and list2:
        if list1.val < list2.val:
            current.next = list1
            list1 = list1.next
        else:
            current.next = list2
            list2 = list2.next
        
        current = current.next
    
    if list1:
        current.next = list1
    if list2:
        current.next = list2
    
    return dummy.next


def printList(node):
    while node:
        print(node.val, end=" ")
        node = node.next
    print()

list1 = ListNode(1)
list1.next = ListNode(2)
list1.next.next = ListNode(4)

list2 = ListNode(1)
list2.next = ListNode(3)
list2.next.next = ListNode(4)

merged_list = merge_two_lists(list1, list2)
printList(merged_list)